README.md
<!-- badges: start -->
[![Travis build status](https://travis-ci.org/Blunde1/covboost.svg?branch=master)](https://travis-ci.org/Blunde1/covboost)
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![GitHub license](https://img.shields.io/badge/license-GPL%20(%3E%3D%202)-blue)](https://www.gnu.org/licenses/gpl-3.0.html)
---------


# covboost
Estimate (possibly sparse, high dimensional) covariance matrix through boosting

`covboost` boosts out the path of the correlation matrix starting from the Identity matrix and eventual convergence (if enough iterations are used) at the ordinary [covariance matrix estimate](https://en.wikipedia.org/wiki/Estimation_of_covariance_matrices#Estimation_in_a_general_context).

## Installation

Install the development version from GitHub
```r
devtools::install_github("Blunde1/covboost")
```
The package is hopefully soon on CRAN.

## Example code and documentation

There are two main functions
- `covboost_cv`: Boosts out the covariance matrix using k-fold cross validation.
- `covboost`: Boosts out the covariance matrix.

The CV function runs in parallel (OpenMP), and returns useful information and plots from cross validation and the best (in terms of Gaussian loss) covariance matrix. See the example below. 
This is information is in the next step used to run `covboost`, which works on the full dataset.

```r
library(covboost)

# -- Generate some Gaussian data with a few (random) non-zero correlations --
p <- 50
n <- 50
x <- matrix(nrow=n, ncol=p)
x[,1] <- rnorm(n)
for(j in 2:p){
    rho <- runif(1,-1,1)*rbinom(1,1,0.5)
    col_dependence <- sample(j-1,1)
    x[,j] <- rnorm(n, mean=rho*x[,col_dependence], sd=(1-rho^2))
}

# -- Run `covboost_cv` to find the best iteration and check convergence --
lrn_rate <- 0.5
cov_cv <- covboost_cv(x, learning_rate=lrn_rate)

# -- Plot to check convergence --
cov_cv$cvplot
cov_cv$opt_iter

# -- Run `covboost` with the best iteration and shrinkage --
sigma <- covboost(x, niter=cov_cv$opt_iter, learning_rate=lrn_rate)
sigma$cov
sigma$cor
sigma$plot # plot (sparse) covariance matrix
```
The code above generates the convergence plot (top), and images of the obtained (sparse) boosted covariance matrix. For reference, here is the ordinary (dense) covariance matrix estimate (bottom right). Note a few significant non-zero correlations in the boosted covariance estimate (bottom left).
<img src="man/figures/readme-fig-1.png" width="70%" />


NEWS.md
------------------------------------------------------------------------
covboost 0.1.0 (2020-04-12)
------------------------------------------------------------------------

- Pre-release.


R/covboost.R
## Copyright (C) 2020 Berent Lunde
## License: GPL-3

#' Boosted Covariance Matrix Estimation
#'
#' @param x An \code{n x p} numeric matrix
#' @param learning_rate Scaling the path of elements in the covariance matrix
#' @param niter The number of boosting iterations
#' @return A list of useful boosting information, most importantly the covariance matrix
#' @examples
#' \dontrun{
#' ## Generate data with random correlation and estimate
#' p <- 50
#' n <- 50
#' x <- matrix(nrow=n, ncol=p)
#' x[,1] <- rnorm(n)
#' for(j in 2:p){
#'     rho <- runif(1,-1,1)*rbinom(1,1,0.5)
#'     col_dependence <- sample(j-1,1)
#'     x[,j] <- rnorm(n, mean=rho*x[,col_dependence], sd=(1-rho^2))
#' }
#' lrn_rate <- 0.2
#' cov_cv <- covboost_cv(x, learning_rate=lrn_rate)
#' cov_cv$cvplot
#' cov_cv$opt_iter
#' sigma <- covboost(x, niter=cov_cv$opt_iter, learning_rate=lrn_rate)
#' sigma$cov
#' sigma$plot
#' }
#' @importFrom stats cov cov2cor quantile
#' @importFrom utils setTxtProgressBar txtProgressBar
#' @rdname covboost
#' @export
covboost <- function(x, learning_rate=0.5, niter=1000)
{
    # Boosts out a covariance matrix from the Identity matrix
    # for p>>n matrix will become singular. The function notices this and terminates
    # Optimal covariance matrix is likely reached some iterations before this

    # requires: mvtnorm, ggplot2

    # x: data
    # learning_rate: shrink each step
    # niter: max number of iterations

    # see CV version for comments on almost identical code

    `10-fold cv` <- cv <- Var1 <- Var2 <- iterations <- qlower <- qupper <- value <- NULL

    # Check input
    error_messages <- c()
    error_messages_type <- c(
        "x" = "\n Error: x must be a numeric matrix",
        "lrn_rate" = "\n Error: learning_rate must be a number between 0 and 1",
        "niter" = "\n Error: niter must be an integer >= 1"
    )

    # check x
    if(!is.matrix(x) || !is.numeric(x)){
        error_messages <- c(error_messages, error_messages_type["x"])
    }

    # learning_rate
    if(!is.numeric(learning_rate) ||
       !(length(learning_rate)==1) ||
       learning_rate < 0 ||
       learning_rate > 1){
        error_messages <- c(error_messages, error_messages_type["lrn_rate"])
    }

    # niter
    if(!is.numeric(niter) ||
       length(niter)>1 ||
       niter < 1){
        error_messages <- c(error_messages, error_messages_type["niter"])
    }

    # Any error messages?
    if(length(error_messages)>0)
        stop(error_messages)


    # Dimensions
    n <- nrow(x)
    p <- ncol(x)

    nll <- numeric(niter)
    stops <- niter+1

    # Matrices
    I <- B_rho <- diag(p)
    A <- cov(x)
    V <- diag(diag(A))

    # Standardization
    A_rho <- cov2cor(A)

    if(niter>1){
        cat("starting boosting...\n")
        pb <- txtProgressBar(min=0, max=niter+1, style=3)
        for(i in 2:(niter+1))
        {
            # max diff
            D_rho <- A_rho - B_rho
            ind <- which(abs(D_rho)==max(abs(D_rho)), arr.ind=T)

            # update
            B_rho[ind] <- B_rho[ind] + learning_rate*D_rho[ind]

            # Does not need check -- I think...!

            setTxtProgressBar(pb, value=i)
        }
        close(pb)
    }

    # transform to final matrix
    B <- round(sqrt(V) %*% B_rho %*% sqrt(V), digits=9)

    if (requireNamespace("ggplot2", quietly = TRUE) && requireNamespace("reshape2", quietly = TRUE)) {
        .cov_heatmap <- function(cov_, title=""){
            # Get upper triangle of the correlation matrix
            get_upper_tri <- function(cormat){
                cormat[lower.tri(cormat)]<- NA
                return(cormat)
            }
            upper_tri <- get_upper_tri(cov_)

            # Melt the correlation matrix
            melted_cormat <- reshape2::melt(upper_tri, na.rm = TRUE)
            # Heatmap
            ggplot2::ggplot(data = melted_cormat, ggplot2::aes(Var2, Var1, fill = value))+
                ggplot2::geom_tile(color = "white")+
                ggplot2::scale_fill_gradient2(low = "blue", high = "red", mid = "white",
                                     midpoint = 0, limit = c(-1,1), space = "Lab",
                                     name="Pearson\nCorrelation") +
                ggplot2::theme_bw()+
                ggplot2::ggtitle(title) +
                ggplot2::theme(
                    legend.justification = c(1, 0),
                    legend.position = c(0.6, 0.7),
                    legend.direction = "horizontal")#+
            #coord_fixed()
        }

        p <- .cov_heatmap(cov2cor(B), "Boosted Correlation Matrix")

    } else {
        p <- cat("Install ggplo2 and reshape2 packages to get cov boosting \n")
    }


    res <- list(
        cov=B,
        cor=B_rho,
        niter=stops,
        plot=p
    )

    return(res)

}


src/dmvnorm.cpp
// Credit to https://gallery.rcpp.org/articles/dmvnorm_arma/
// Modified by Berent Lunde

#include <RcppArmadillo.h>
#include <omp.h>

// [[Rcpp::depends("RcppArmadillo")]]

// [[Rcpp::plugins(openmp)]]


static double const log2pi = std::log(2.0 * M_PI);

void inplace_tri_mat_mult(arma::rowvec &x, arma::mat const &trimat){
    arma::uword const n = trimat.n_cols;

    for(unsigned j = n; j-- > 0;){
        double tmp(0.);
        for(unsigned i = 0; i <= j; ++i)
            tmp += trimat.at(i, j) * x[i];
        x[j] = tmp;
    }
}

// [[Rcpp::export(.dmvnorm_arma_mc)]]
arma::vec dmvnrm_arma_mc(arma::mat const &x,
                         arma::rowvec const &mean,
                         arma::mat const &sigma,
                         bool const logd = false,
                         int const cores = 1) {
    using arma::uword;
    omp_set_num_threads(cores);
    uword const n = x.n_rows,
        xdim = x.n_cols;
    arma::vec out(n);

    try{
        arma::mat const rooti = arma::inv(trimatu(arma::chol(sigma)));
        double const rootisum = arma::sum(log(rooti.diag())),
            constants = -(double)xdim/2.0 * log2pi,
            other_terms = rootisum + constants;

        arma::rowvec z;
        #pragma omp parallel for schedule(static) private(z)
        for (uword i = 0; i < n; i++) {
            z = (x.row(i) - mean);
            inplace_tri_mat_mult(z, rooti);
            out(i) = other_terms - 0.5 * arma::dot(z, z);
        }

        if (logd)
            return out;
        return exp(out);
    }  catch(std::exception &ex) {
        forward_exception_to_r(ex);
    } catch(...) {
        Rcpp::stop("c++ exception (unknown reason)");
    }

    out.fill(NA_REAL);
    return out;             // not reached


}


src/RcppExports.cpp
// Generated by using Rcpp::compileAttributes() -> do not edit by hand
// Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#include <RcppArmadillo.h>
#include <Rcpp.h>

using namespace Rcpp;

// dmvnrm_arma_mc
arma::vec dmvnrm_arma_mc(arma::mat const& x, arma::rowvec const& mean, arma::mat const& sigma, bool const logd, int const cores);
RcppExport SEXP _covboost_dmvnrm_arma_mc(SEXP xSEXP, SEXP meanSEXP, SEXP sigmaSEXP, SEXP logdSEXP, SEXP coresSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< arma::mat const& >::type x(xSEXP);
    Rcpp::traits::input_parameter< arma::rowvec const& >::type mean(meanSEXP);
    Rcpp::traits::input_parameter< arma::mat const& >::type sigma(sigmaSEXP);
    Rcpp::traits::input_parameter< bool const >::type logd(logdSEXP);
    Rcpp::traits::input_parameter< int const >::type cores(coresSEXP);
    rcpp_result_gen = Rcpp::wrap(dmvnrm_arma_mc(x, mean, sigma, logd, cores));
    return rcpp_result_gen;
END_RCPP
}

static const R_CallMethodDef CallEntries[] = {
    {"_covboost_dmvnrm_arma_mc", (DL_FUNC) &_covboost_dmvnrm_arma_mc, 5},
    {NULL, NULL, 0}
};

RcppExport void R_init_covboost(DllInfo *dll) {
    R_registerRoutines(dll, NULL, CallEntries, NULL, NULL);
    R_useDynamicSymbols(dll, FALSE);
}


R/covboost_cv.R
## Copyright (C) 2020 Berent Lunde
## License: GPL-3

#' Boosted Covariance Matrix Estimation With CV
#'
#' @param x An \code{n x p} numeric matrix
#' @param learning_rate Scaling the path of elements in the covariance matrix
#' @param niter The number of boosting iterations
#' @param nfolds The number of folds for k-fold cross validation
#' @param cores The number of cores use for parallel computations
#'
#' @return A list of useful cross validation information from boosting out the covariance matrix
#' @examples
#' \dontrun{
#' ## Generate data with random correlation
#' p <- 50
#' n <- 50
#' x <- matrix(nrow=n, ncol=p)
#' x[,1] <- rnorm(n)
#' for(j in 2:p){
#'     rho <- runif(1,-1,1)*rbinom(1,1,0.5)
#'     col_dependence <- sample(j-1,1)
#'     x[,j] <- rnorm(n, mean=rho*x[,col_dependence], sd=(1-rho^2))
#' }
#' lrn_rate <- 0.5
#' cov_cv <- covboost_cv(x, learning_rate=lrn_rate)
#' cov_cv$cvplot
#' cov_cv$opt_iter
#' }
#'
#' @importFrom stats cov cov2cor quantile
#' @importFrom utils setTxtProgressBar txtProgressBar
#' @rdname covboost_cv
#' @export
covboost_cv <- function(x, learning_rate=0.5, niter=1000, nfolds=10, cores=1)
{
    # Boosts out a covariance matrix from the Identity matrix
    # for p>>n matrix will become singular. The function notices this and terminates
    # Optimal covariance matrix is likely reached some iterations before this

    # requires: ggplot2

    # x: data
    # learning_rate: shrink each step
    # niter: max number of iterations
    # nfolds: number of folds in CV
    # cores: number of cores used in parallel

    `10-fold cv` <- cv <- Var1 <- Var2 <- iterations <- qlower <- qupper <- value <- NULL

    # Check input
    error_messages <- c()
    error_messages_type <- c(
        "x" = "\n Error: x must be a numeric matrix",
        "lrn_rate" = "\n Error: learning_rate must be a number between 0 and 1",
        "niter" = "\n Error: niter must be an integer >= 1",
        "nfolds" = "\n Error: nfolds must be an integer > 1 and <= nrow(x)",
        "cores" = "\n Error: cores must be an integer >= 1"
    )

    # check x
    if(!is.matrix(x) || !is.numeric(x)){
        error_messages <- c(error_messages, error_messages_type["x"])
    }

    # learning_rate
    if(!is.numeric(learning_rate) ||
       !(length(learning_rate)==1) ||
       learning_rate < 0 ||
       learning_rate > 1){
        error_messages <- c(error_messages, error_messages_type["lrn_rate"])
    }

    # niter
    if(!is.numeric(niter) ||
       length(niter)>1 ||
       niter < 1){
        error_messages <- c(error_messages, error_messages_type["niter"])
    }

    # nfolds
    if(!is.numeric(nfolds) ||
       length(nfolds)>1 ||
       nfolds < 1 ||
       nfolds > nrow(x)){
        error_messages <- c(error_messages, error_messages_type["nfolds"])
    }

    # cores
    if(!is.numeric(cores) ||
       length(cores)>1 ||
       cores < 1){
        error_messages <- c(error_messages, error_messages_type["cores"])
    }

    # Any error messages?
    if(length(error_messages)>0)
        stop(error_messages)


    # Dimensions
    n <- nrow(x)
    p <- ncol(x)

    # Set up variables
    MAX_CONSECUTIVE_NO_IMPROVEMENT <- 50
    NO_IMPROVEMENT_COUNTER <- 0
    BEST <- NULL
    cv_nll_cor_k <- numeric(niter+1)
    cv_nll_cor <- matrix(nrow=niter+1, ncol=nfolds)
    stops <- rep(niter+1, nfolds)

    pb <- txtProgressBar(min=0, max=(niter+1)*nfolds, style=3)
    for(k in 1:nfolds)
    {
        # reset counter
        NO_IMPROVEMENT_COUNTER <- 0

        # define holdout set
        options(warn=-1)
        holdout <- split(sample(n,n), 1:nfolds)
        options(warn=0)

        # starting matrices
        Ak <- cov(x[-holdout[[k]],])
        Vk <- diag(diag(Ak))
        eDiag <- sqrt(Vk)
        invDiag <- sqrt(diag(1/diag(Vk)))
        Ak_rho <- invDiag %*% Ak %*% invDiag
        Bk_rho <- diag(p) # Identity: inverse of diagonal variance matrix
        Bk <- round(eDiag %*% Bk_rho %*% eDiag, digits=9) # = Vk = diag(diag(Ak))
        cv_nll_cor_k[1] <- NA
        cv_nll_cor_k[1] <- try(-sum(.dmvnorm_arma_mc(x[holdout[[k]],], rep(0,p), Bk, logd = TRUE, cores=cores)), silent = T)
        BEST <- cv_nll_cor_k[1]

        for(i in 2:(niter+1))
        {
            # difference
            Dk_rho <- Ak_rho - Bk_rho
            # COOL NOTE: POSSIBLE TO HELP WITH GRAPH!!
            ind <- which(abs(Dk_rho)==max(abs(Dk_rho)), arr.ind=T)

            # update Bk_rho
            Bk_rho[ind] <- Bk_rho[ind] + learning_rate*Dk_rho[ind]

            # check holdout loss
            Bk <- round(eDiag %*% Bk_rho %*% eDiag, digits=9) # transform to correlation, round ensures symmetry
            cvnll_i_k <- NA # default if fails
            cvnll_i_k <- try(-sum(.dmvnorm_arma_mc(x[holdout[[k]],], rep(0,p), Bk, logd = TRUE, cores=cores)), silent = T)

            # checks
            if(!is.finite(cvnll_i_k)) {
                # fill remaining
                cv_nll_cor_k[i:(niter+1)] <- cv_nll_cor_k[i-1]
                # get stop-point
                stops[k] <- i-1
                #cat("stopping at iteration: ", i, "\n",
                #    "updating niter to ", i-1)
                break
            }else{
                #update
                cv_nll_cor_k[i] <- cvnll_i_k

                # check counter
                if(cv_nll_cor_k[i]>=BEST){
                    NO_IMPROVEMENT_COUNTER <- NO_IMPROVEMENT_COUNTER + 1
                }else{
                    NO_IMPROVEMENT_COUNTER <- 0
                }
            }

            # check counter
            if(NO_IMPROVEMENT_COUNTER==MAX_CONSECUTIVE_NO_IMPROVEMENT){
                cv_nll_cor_k[i:(niter+1)] <- cv_nll_cor_k[i-1]
                stops[k] <- i
                break
            }

            setTxtProgressBar(pb, value=(k-1)*niter+i)
        }

        cv_nll_cor[,k] <- cv_nll_cor_k

    }
    close(pb)

    cat("preparing data...\n")
    cv_nll <- cv_nll_cor[1:max(stops),,drop=FALSE]
    cv_nll_mean <- rowMeans(cv_nll, na.rm=T)
    cv_nll_qupper <- sapply(1:nrow(cv_nll), function(i){quantile(cv_nll[i,], 0.6, na.rm = T)})
    cv_nll_qlower <- sapply(1:nrow(cv_nll), function(i){quantile(cv_nll[i,], 0.4, na.rm=T)})
    chol_decomp_fails <- stops[which(stops < (niter+1))]

    # update niter and create data
    niter <- nrow(cv_nll)
    .df <- data.frame(1:niter, cv_nll_mean[1:niter])
    .df2 <- data.frame(1:niter, cv_nll_qupper[1:niter], cv_nll_qlower[1:niter])
    names(.df) <- c("iterations", "10-fold cv")
    names(.df2) <- c("iterations", "qupper", "qlower")

    # CV loss vs iterations plot
    cat("preparing loss...\n")
    cbp2 <- c("#000000", "#E69F00", "#56B4E9", "#009E73",
              "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
    names(cbp2) <- colnames(.df)
    opt_iter = .df[which.min(.df$`10-fold cv`), "iterations"]

    if (requireNamespace("ggplot2", quietly = TRUE)) {
        .p <- ggplot2::ggplot(data=.df) +
            ggplot2::geom_point(ggplot2::aes(x=iterations, y=`10-fold cv`), colour=cbp2[2]) +
            ggplot2::geom_line(ggplot2::aes(x=iterations, y=`10-fold cv`), colour=cbp2[2]) +
            ggplot2::geom_ribbon(data=.df2, ggplot2::aes(iterations, ymax=qupper, ymin=qlower), fill=cbp2[2], alpha=0.5) +
            ggplot2::geom_vline(xintercept=opt_iter, size=1) +
            ggplot2::xlab("Iteration") +
            ggplot2::ylab("Gaussian loss") +
            ggplot2::ggtitle("Gaussian CV Loss Versus Iterations") +
            ggplot2::theme_bw() +
            ggplot2::geom_vline(xintercept=chol_decomp_fails, size=1, colour="blue")
    } else {
        .p <- cat("Install ggplo2 package to get cv-plots \n")
    }

    cat("preparing results...\n")
    .res_data <- data.frame(1:niter, cv_nll_mean[1:niter], cv_nll_qupper[1:niter], cv_nll_qlower[1:niter])
    names(.res_data) <- c("iterations", "cv-mean", "cv_6_quantile", "cv_4_quantile")
    res <- list(
        #opt_shrinkage = d_min_loss,
        opt_iter = opt_iter,
        cvplot = .p,
        data = .res_data,
        chol_decomp_fails = chol_decomp_fails
    )


    return(res)

}
